src/main/java/ml/shifu/shifu/core/dtrain/dt/DTEarlyStopDecider.java

```diff
@@ -0,0 +1,307 @@
+            }
+        };
+
+        validationErrorDecider = new MinAverageDecider(this.treeDepth, this.treeDepth * this.magicNumber) {
+            @Override
+            public boolean getDecide() {
+                return this.gain < 0;
+            }
+        };
+    }
+
+    /**
+     * Add new iteration's train error and validation error into the decider.
+     *
+     * @param trainError training error
+     * @param validationError validation error
+     * @return true if no more iteration needed, else false
+     */
+    public boolean add(double trainError, double validationError) {
+        boolean trainErrorDecideReady = this.trainErrorDecider.add(trainError);
+        boolean validationDecideReady = this.validationErrorDecider.add(validationError);
+
+        if (trainErrorDecideReady) {
+            if(trainErrorDecider.getDecide()){
+                this.trainGainContinueLowerCount += 1;
+                LOG.warn("Continue " + this.trainGainContinueLowerCount + " positive sign for not worth more iteration!");
+            } else {
+                this.trainGainContinueLowerCount = 0;
+            }
+        }
+
+        if (validationDecideReady) {
+            if(validationErrorDecider.getDecide()){
+                this.validationGainContinueNegCount += 1;
+                LOG.warn("Continue " + this.validationGainContinueNegCount + " positive sign for not worth more iteration!");
+            } else {
+                this.validationGainContinueNegCount = 0;
+            }
+        }
+
+        return canStop();
+    }
+
+    private boolean canStop() {
+        if (this.validationGainContinueNegCount >= magicNumber) {
+            return true;
+        }
+        if (this.trainGainContinueLowerCount >= magicNumber) {
+            return true;
+        }
+        return false;
+    }
+
+    abstract class MinAverageDecider {
+        /**
+         * minQueue to get the minimal value of a queue size values
+         */
+        private final ThreadLocal<MinQueue> minQueue;
+        /**
+         * averageQueue, insert with recursive average value into the queue, and get iteration gain
+         */
+        private AverageQueue averageQueue;
+        /**
+         * Max gain so far
+         */
+        protected double maxGain;
+        /**
+         * Current gain
+         */
+        protected double gain;
+
+        MinAverageDecider(int minQueueNum, int averageQueueNum) {
+            minQueue = new ThreadLocal<MinQueue>();
+            this.minQueue.set(new MinQueue(minQueueNum));
+            this.averageQueue = new AverageQueue(averageQueueNum);
+            this.maxGain = 0;
+        }
+
+        /**
+         * Add a value into the decider
+         * @param element the value to insert to the decide
+         * @return true if new gain generated, and decide is ready to get
+         */
+        public boolean add(double element) {
+            if (this.minQueue.get().add(element) == false) {
```

Comment:
this.minQueue.get().add(element) == false
or
!(this.minQueue.get().add(element))